### ISSUE TITLE
Limit the number of audio-chunks waiting to be transcribed

### ISSUE THREAD

**Issue by sergutsan** [fernandezgom/it2l-platform#76]
At the moment, every time the platform receives an audio chunk, a new thread is created to call SAIL's transcription service and receive some transcribed words in return. 

Problem: if SAIL does not respond immediately, the thread keeps waiting and using 500kB of memory. After a few seconds, this starts to be a lot of memory wasted (and increasing). 

A solution has two parts: 

a) Make all calls (platform - orchestrator - speech module - SAIL) asynchronous except the last one ( - SAIL). This can be done by callbacks or by opening getNewWords(String[]) services on platform and orchestrator. 

b) Implement a semaphore to limit the amount of threads that can be waiting for SAIL to respond. After a certain threshold has been reached (3? 5? 10?), no more threads can call SAIL and their audio-chunk is thus discarded. This will mean losing words, but makes the system much more robust against SAIL's unresponsiveness.

**Comment by sergutsan** [comment]
For additional context / details, here is a former email interchange...

...

Problem: sometimes a speech-reco instance becomes unresponsive and
that results not only in no transcriptions, but in a lot of memory
wasted as threads are kept on hold.

My solution has two parts: (a) using semaphores and (b) wasting cheap
memory (meaning on the windows machines).

a. Semaphores

This is the easiest part but it is also the most important, so IMHO it
should be implemented straight away.

I think we should implement semaphores to access the speech-reco
instances. As you know, a semaphore is basically a counter of
processes using a resource. When too many processes are using said
resource, no more processes are allowed to use it until at least one
of the current processes releases their use. A mutex lock (e.g. Java´s
keyword "synchronized") is just a semaphore of size 1: only one thread
can come into a synchronized portion of code, the others must wait
until the current thread releases the lock.

In the case of speech-reco we must make a decision of how many threads
we will allow to be waiting for speech-reco to answer. I would say
something between three and five, but I will be happy even with one.
The pseudocode for each thread would look like:

1. Collect 0.5 MB of data
2. If semaphore is positive.
  3a. Reduce semaphore by one
  4a. Send data to SAIL
  5a. Receive words from SAIL.
  6a. Discard data
  7a. Increase semaphore by one
  else
  3a. Discard data
  4a. Words to return are thus empty string
end_if
8. Return words to interested parties

This means that we will be losing words, but I prefer to lose words
from a lousy transcription that to make the system crash.

Does this make sense?



b. Wasting Windows memory

I think the latest call to SAIL should be from a process on the
Windows machine, not on the linux machine.

This allows the threads on the linux machine to return quickly and to
discard the memory quickly. Regardless of how many threads are kept
waiting by SAIL (and we control this with the size of our semaphore,
see above) they will never crash or hang the linux machine, so
everything else in iTalk2Learn will work fine (assuming CTAT behaves).
The worst-case-scenario is that we do not get any transcriptions,
which is acceptable, instead of the current worst-case-scenario in
which the system becomes unresponsive or freezes and requires
rebooting.

The data flow would be something like:

1. Web browser on client machine records voice (0.5MB)

2. Client sends data to server (iTalk2learn linux machine)

3. iTalk2Learn server sends data to orchestrator, including an ID for
sorting purposes. This is a call that returns immediately. Memory is
not released yet because orchestrator has pointer to it.

4. Assuming semaphore is open, orchestrator sends data to webservice
on windows machine, including ID. This is a call that returns
immediately. Memory is released. If semaphore is closed, data is
discarded and memory is also released.

5. Webservice on windows calls SAIL (on same windows machine). This
call only returns when SAIL provides an answer (maybe including
transcribed words) so it may hang forever.

6. When SAIL returns words, those words are sent back to the
orchestrator via a callback (or maybe a dedicated service) together
with the ID (so they can be put in place back in the server even if
SAIL discarded data or answered in a non-ordered fashion).

**Comment by fernandezgom** [comment]
Done in both sides. Asynchronous process in the client side and semaphore controls one minute of inactivity of Speech Reco with a counter not to send more audio.

**Comment by sergutsan** [comment]
Can you please provide a link to the relevant commit? :-)

I think you can add "Fixes Issue 76" (or something of the sort) on the commit message and a link is created automagically.

**Comment by fernandezgom** [comment]
https://github.com/fernandezgom/it2l-speech-orchestration-module
https://github.com/fernandezgom/it2l-speech-module
